src/Pact/Types/Advice.hs

```diff
@@ -0,0 +1,83 @@
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+-- |
+-- Module      :  Pact.Types.Advice
+-- Copyright   :  (C) 2020 Stuart Popejoy
+-- License     :  BSD-style (see the file LICENSE)
+-- Maintainer  :  Stuart Popejoy <stuart@kadena.io>
+--
+-- Advise user functions in runtime for performance, coverage etc.
+--
+
+module Pact.Types.Advice
+  ( Advice(..)
+  , AdviceContext(..)
+  , advise
+  , DbContext(..)
+  ) where
+
+import Control.Lens
+import Control.Monad.IO.Class
+import Data.Default
+import Data.Text (Text)
+import Pact.Types.Hash
+import Pact.Types.Info
+import Pact.Types.Term
+
+-- | Enum indicating which Db operation is running.
+data DbContext =
+    DbRead
+  | DbWrite
+  | DbKeys
+  | DbTxIds
+  | DbCreateUserTable
+  | DbGetUserTableInfo
+  | DbBeginTx
+  | DbCommitTx
+  | DbRollbackTx
+  | DbGetTxLog
+  deriving (Eq,Show,Enum,Bounded)
+
+-- | Type of advised operation.
+data AdviceContext =
+    AdviceUser (Def Ref,[Term Name])
+    -- ^ Advise on user function
+  | AdviceNative (NativeDefName)
+    -- ^ Advise on native
+  | AdviceTx PactHash
+    -- ^ Transaction execution wrapper
+  | AdviceDb DbContext
+    -- ^ Db operation
+  | AdviceOther Text
+    -- ^ Arbitrary advice.
+  deriving (Eq,Show)
+
+
+-- | Bracket some Pact operation.
+newtype Advice = Advice {
+  _advise :: forall m a . MonadIO m
+    => Info
+    -> AdviceContext
+    -> m a
+    -> m a
+  }
+
+instance Default Advice where def = Advice defAdvice
+instance Show Advice where show _ = "Advice"
+instance Semigroup Advice where
+  Advice f <> Advice g =
+    Advice $ \i ctx act -> f i ctx $! g i ctx act
```

Comment:
Will the order of application change the state of the database and the outcome of the functions here? If so, we might not want to expose a semigroup instance and just defer to the underlying monad instead which is the composition.

Reply:
It shouldn't? The intent here is to have `f` bracket `g` bracket `act`, and `act` is the actual side-effect. `f` and `g` have no access to the `Eval` monad so while they could buggily call the action twice (bad, but a bug), I don't think the excess bracketing changes any outside ordering.